File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <math.h>

using namespace std;

int dfs(int cur, vector<bool>& visited, vector<vector<int>>& adj,
       int from, int to) {
    int ret = 1;
    visited[cur] = true;
    
    for(int next : adj[cur]){
        if((cur == from && next == to) || (cur == to && next == from))
            continue;
        
        if(!visited[next]){
            ret += dfs(next, visited, adj, from, to);
        }
    }
    return ret;
}

int solution(int n, vector<vector<int>> wires) {
    
    // dfs
    vector<vector<int>> adj(n + 1);
    
    for(vector<int> wire : wires) {
        adj[wire[0]].push_back(wire[1]);
        adj[wire[1]].push_back(wire[0]);
    }
    
    int ret = 987654321;
    for(vector<int> wire : wires) {
        vector<bool> visited(n + 1, false); // 1 ~ n개
        int from = wire[0]; int to = wire[1];
        vector<int> result;
        for(int i = 1 ; i <= n ; i++)
            if(!visited[i])
                result.push_back(dfs(i, visited, adj, from, to));
        ret = min(ret, abs(result[0] - result[1]));
    }
    
    return ret;
    
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;
import java.util.stream.*;

class Solution {
    public int solution(int n, int[][] wires) {
        int min = Integer.MAX_VALUE;
        Map<Integer, List<Integer>> map = new HashMap<>();
        for(int i = 0 ; i < wires.length ; i++){
            int a = wires[i][0];
            int b = wires[i][1];
            map.computeIfAbsent(a, k -> new ArrayList<>()).add(b);
            map.computeIfAbsent(b, k -> new ArrayList<>()).add(a);
        }
        
        // 하나씩 끊어 보기
        for(int[] wire : wires){
            int a = wire[0];
            int b = wire[1];
            
            map.get(a).remove((Integer) b);
            map.get(b).remove((Integer) a);
            
            int aCnt = dfs(map, a, new boolean[n + 1]);
            int bCnt = n - aCnt;
            
            min = Math.min(min, Math.abs(aCnt - bCnt));
            
            map.get(a).add(b);
            map.get(b).add(a);
        }
        
        return min;
    }
    
    int dfs(Map<Integer, List<Integer>> map, int cur, boolean[] visited){
        
        int cnt = 1;
        visited[cur] = true;
        
        for(int next : map.get(cur)){
            if(!visited[next]){
                cnt += dfs(map, next, visited);
            }
        }
        
        return cnt;
    }
}